substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeBootImage.java

```diff
@@ -952,4 +973,299 @@ protected NativeTextSectionImpl(RelocatableBuffer relocatableBuffer, ObjectFile
         protected final ObjectFile objectFile;
         protected final NativeImageCodeCache codeCache;
     }
+
+    /**
+     * implementation of the DebugInfoProvider API interface
+     * that allows type, code and heap data info to be passed to
+     * an ObjectFile when generation of debug info is enabled.
+     */
+    private class NativeImageDebugInfoProvider implements DebugInfoProvider {
+        private final NativeImageCodeCache codeCache;
+        private final NativeImageHeap heap;
+        private final Iterator<Map.Entry<HostedMethod, CompilationResult>> codeCacheIterator;
+        private final Iterator<Map.Entry<Object, ObjectInfo>> heapIterator;
+
+        NativeImageDebugInfoProvider(NativeImageCodeCache codeCache, NativeImageHeap heap) {
+            super();
+            this.codeCache = codeCache;
+            this.heap = heap;
+            this.codeCacheIterator = codeCache.compilations.entrySet().iterator();
+            this.heapIterator = heap.objects.entrySet().iterator();
+        }
+
+        @Override
+        public DebugTypeInfoProvider typeInfoProvider() {
+            return () -> new Iterator<DebugTypeInfo>() {
+                @Override
+                public boolean hasNext() {
+                    return false;
+                }
+
+                @Override
+                public DebugTypeInfo next() {
+                    return null;
+                }
+            };
+        }
+
+        @Override
+        public DebugCodeInfoProvider codeInfoProvider() {
```

Comment:
afaics this is the only thing used atm, right?

Reply:
Yes, at present. However, the plan is to also allow the debug writer to iterate over types and details of heap contents courtesy of a DebugTypeInfoProvider and DebugDataInfoProvider. The details of those interfaces still need to be fully worked out.